Add ORM

jjnesbitt opened the pull request:
Fixes multinet-app/multinet-server#442.

This adds an ORM for database operations, and better logical divisions between workspaces, tables, graphs, and users. Notably, this PR **does not** address the temporary permissions transforms being done when interfacing with the permissions endpoints. I held off on this because this is a big enough change as it is, and if I also included that change, it would require a data migration. I've created multinet-app/multinet-server#456 to track this.

All tests are passing (at least locally for me). Please take your time as there's a lot to unpack here.

AlmightyYakob commented:
Also just a note, one of my biggest gripes with the current implementation is the sheer amount of methods on the `Workspace` class. Ideally these would be more split up between the classes, but I'm not sure if there's a sane way to do it. If you guys come up with any suggestions regarding this, please lmk.

jjnesbitt committed:
Add Graph class

jjnesbitt committed:
Add Table class

jjnesbitt committed:
Add Workspace class

jjnesbitt committed:
Add system_db func, add readonly flag to db funcs

jjnesbitt committed:
Replace user functions with User class, fix types

jjnesbitt committed:
Update auth to use User and Workspace classes

jjnesbitt committed:
Use Workspace/Graph/Table classes in downloaders

jjnesbitt committed:
Update uploaders to use ORM classes

jjnesbitt committed:
Remove old db functions, add two user funcs

- Add the `user_collection` function to this file.
- Add the `search_user` function to this file

These functions are placed here instead of `user.py` because they are
simple low-level database operations, and are separated from the User
model.

jjnesbitt committed:
Move permission transforms into util.py

jjnesbitt committed:
Replace db funcs with ORM class methods

jjnesbitt committed:
Add missing mypy stubs

jjnesbitt committed:
Update tests to use ORM classes

JackWilb commented on multinet/workspace.py:
```diff
@@ -0,0 +1,356 @@
+    @staticmethod
+    def from_dict(d: Dict) -> Workspace:
+        """Construct a workspace from a dict."""
+        workspace = Workspace(name=d["name"])
```
We could check if `d["name"]` exists first and throw InternalServerError if not

AlmightyYakob replied:
Would be covered under https://github.com/multinet-app/multinet-server/issues/460

JackWilb commented on multinet/workspace.py:
```diff
@@ -0,0 +1,356 @@
+
+        permissions = d.get("permissions")
+        if permissions:
+            workspace.permissions = WorkspacePermissions(**permissions)
```
```suggestion
            workspace.set_permissions(WorkspacePermissions(**permissions))
```

AlmightyYakob replied:
This code was actually redundant/incorrect, as this is all handled in the constructor of `Workspace`, and has been removed in b93c5ea

JackWilb commented on multinet/uploaders/d3_json.py:
```diff
@@ -107,28 +108,20 @@ def upload(workspace: str, graph: str) -> Any:
+    if loaded_workspace.has_table(node_table_name):
+        node_table = loaded_workspace.table(node_table_name)
     else:
-        nodes_coll = space.create_collection(node_table_name, edge=False)
+        node_table = loaded_workspace.create_table(node_table_name, edge=False)
 
-    if space.has_collection(edge_table_name):
-        links_coll = space.collection(edge_table_name)
+    if loaded_workspace.has_table(edge_table_name):
+        edge_table = loaded_workspace.table(edge_table_name)
     else:
-        links_coll = space.create_collection(edge_table_name, edge=True)
+        edge_table = loaded_workspace.create_table(edge_table_name, edge=True)
```
This seems like it could be buggy. Should we reject the upload if there is already a table with a name that we need?

waxlamp replied:
If that's a bug, it's independent of this PR. Could you file an issue please?

waxlamp replied:
Opened #458.

JackWilb commented on multinet/uploaders/nested_json.py:
```diff
@@ -98,38 +99,31 @@ def upload(workspace: str, graph: str) -> Any:
-        edgetable = space.create_collection(edgetable_name, edge=True)
+        edgetable = loaded_workspace.create_table(edgetable_name, edge=True)
 
-    if space.has_collection(int_nodetable_name):
-        int_nodetable = space.collection(int_nodetable_name)
+    if loaded_workspace.has_table(int_nodetable_name):
+        int_nodetable = loaded_workspace.table(int_nodetable_name)
     else:
-        int_nodetable = space.create_collection(int_nodetable_name)
+        int_nodetable = loaded_workspace.create_table(int_nodetable_name, edge=False)
 
-    if space.has_collection(leaf_nodetable_name):
-        leaf_nodetable = space.collection(leaf_nodetable_name)
+    if loaded_workspace.has_table(leaf_nodetable_name):
+        leaf_nodetable = loaded_workspace.table(leaf_nodetable_name)
     else:
-        leaf_nodetable = space.create_collection(leaf_nodetable_name)
+        leaf_nodetable = loaded_workspace.create_table(leaf_nodetable_name, edge=False)
```
Again, this seems like it will cause us headaches with tables not containing what we expect. We should consider rejecting the upload if the tables exist.

waxlamp replied:
Opened #458.

JackWilb commented on multinet/uploaders/d3_json.py:
```diff
@@ -107,28 +108,20 @@ def upload(workspace: str, graph: str) -> Any:
         del link["target"]
 
     # Create or retrieve the workspace
```
This comment is wrong

JackWilb commented on multinet/uploaders/newick.py:
```diff
@@ -86,17 +87,17 @@ def upload(workspace: str, graph: str) -> Any:
+    if loaded_workspace.has_table(edgetable_name):
+        edgetable = loaded_workspace.table(edgetable_name)
     else:
         # Note that edge=True must be set or the _from and _to keys
         # will be ignored below.
-        edgetable = space.create_collection(edgetable_name, edge=True)
+        edgetable = loaded_workspace.create_table(edgetable_name, edge=True)
 
-    if space.has_collection(nodetable_name):
-        nodetable = space.collection(nodetable_name)
+    if loaded_workspace.has_table(nodetable_name):
+        nodetable = loaded_workspace.table(nodetable_name)
     else:
-        nodetable = space.create_collection(nodetable_name)
+        nodetable = loaded_workspace.create_table(nodetable_name, edge=False)
```
Consider rejecting upload instead

waxlamp replied:
Opened #458.

JackWilb commented on multinet/uploaders/newick.py:
```diff
@@ -105,29 +106,25 @@ def read_tree(parent: Optional[str], node: newick.Node) -> None:
-                }
+                [
+                    {
+                        "_from": "%s/%s" % (nodetable_name, parent),
+                        "_to": "%s/%s" % (nodetable_name, key),
```
```suggestion
                        "_from": f"{nodetable_name}/{parent}",
                        "_to": f"{nodetable_name}/{key}",
```

JackWilb commented:
Looks great so far, Jake! That commit where you remove 500 lines of code felt really nice

waxlamp commented on multinet/db.py:
```diff
@@ -58,42 +58,41 @@
 
 
-def db(name: str) -> StandardDatabase:
+def db(name: str, readonly: bool = True, immediate: bool = False) -> StandardDatabase:
```
What does `immediate` do?

AlmightyYakob replied:
Initializes the connection to the database immediately by sending a test request. This is generally not desired, as it means more upfront requests when we may not need them.

waxlamp replied:
Why did you include it? Do we require immediate access somewhere?

AlmightyYakob replied:
Not sure, I can remove it.

waxlamp commented on multinet/db.py:
```diff
@@ -58,42 +58,41 @@
+
+    username = "readonly" if readonly else "root"
+    password = (
+        os.environ.get("ARANGO_READONLY_PASSWORD", "letmein")
+        if readonly
+        else os.environ.get("ARANGO_PASSWORD", "letmein")
```
You can maybe factor out the if-else expression to apply to just the environment variable.

AlmightyYakob replied:
While true, I personally think it would make the code more messy.

waxlamp replied:
Ok.

waxlamp commented on multinet/db.py:
```diff
@@ -58,42 +58,41 @@
+    return db("_system", readonly)
 
 
+# TODO: Fix/remove this
```
What needs to be checked/removed? If appropriate please file an issue and link here, or remove the comment.

waxlamp commented on multinet/db.py:
```diff
@@ -616,10 +615,10 @@ def count_text(filt: str) -> str:
     if not sysdb.has_database("uploads"):
         sysdb.create_database("uploads")
-    return db("uploads")
+    return db("uploads", readonly=False)
```
The function should take a `readonly` parameter, like `workspace_mapping_collection()` does.

waxlamp commented on multinet/db.py:
```diff
@@ -656,3 +655,4 @@ def delete_upload_collection(upload_id: str) -> str:
     uploads_db.delete_collection(upload_id)
 
     return upload_id
+# TODO: Refactor the below functions into an `Upload` class
```
What does this comment refer to?

AlmightyYakob replied:
It refers to the upload functions from Mike V's PR. It must have got split apart from his changes, but it's visible in the final change.

waxlamp replied:
I did notice this after I left the comment. Oops.

waxlamp replied:
Can you file an issue and link it in the comment?

waxlamp commented on multinet/user.py:
```diff
@@ -1,163 +1,204 @@
 """User data and functions."""
+from __future__ import annotations  # noqa: T484
```
What's this all about?

AlmightyYakob replied:
If you have a class method that returns an instance of that class, and you want to type annotate it properly, this is required. In this case, I believe they're all static methods. Here's an example https://github.com/multinet-app/multinet-server/blob/550d608f6d0dc7452d35672b4edb3d94ea3969e0/multinet/user.py#L92-L100

waxlamp commented on multinet/user.py:
```diff
@@ -1,163 +1,204 @@
+        return User.get(sub) is not None
+
+    @staticmethod
+    def get(sub: str) -> Optional[Dict]:
```
Use a more specific type than `Dict`.

waxlamp replied:
#460

waxlamp commented on multinet/user.py:
```diff
@@ -1,163 +1,204 @@
-    """Return the collection that contains user documents."""
-    sysdb = db("_system")
+@dataclass
+class MultinetInfo:
```
I feel like this needs a much better class name.

Why can't this information be included in the `UserInfo` class directly?

AlmightyYakob replied:
`UserInfo` represents the data that can be used to create a new User, while `MultinetInfo` represents the additional data that is added by our system, and that we use. This may grow, which is why having a separate dataclass to represent it is beneficial.

waxlamp replied:
Makes sense. I still think we should come up with a better name than "MultinetInfo". But I see that you're simply updating what was there before, so I've filed #461 and will resolve this conversation.

waxlamp commented on multinet/user.py:
```diff
@@ -1,163 +1,204 @@
+            return None
+
+    @staticmethod
+    def from_dict(d: Dict) -> User:
```
More specific type than `Dict`.

waxlamp replied:
#460

waxlamp commented on multinet/auth/util.py:
```diff
@@ -27,23 +25,23 @@ def wrapper(*args: Any, **kwargs: Any) -> Any:
     # A non-logged-in user, by definition, is a reader of public workspaces.
     if user is None:
-        return perms["public"]
+        return bool(perms.public)
```
Isn't `perms.public` already a `bool`?

AlmightyYakob replied:
Good point, I'll change it.

waxlamp commented on multinet/downloaders/csv.py:
```diff
@@ -31,13 +26,15 @@ def download(workspace: str, table: str) -> Any:
-    limit = workspace_table_row_count(workspace, table)
-    table_rows = workspace_table_rows(workspace, table, 0, limit)
-    fields = workspace_table_keys(workspace, table, filter_keys=True)
+    loaded_table = loaded_workspace.table(table)
+    limit = loaded_table.row_count()
+    table_rows = loaded_table.rows(0, limit)["rows"]
```
Can't you just call `.rows()` without the `limit` parameter and get all the rows? Having to call `.row_count()` first seems like bad design.

AlmightyYakob replied:
Yeah, I'll make this change. This is vestigial from the previous implementation, not sure why I didn't notice this.

waxlamp commented on multinet/db.py:
```diff
@@ -132,358 +107,33 @@ def workspace_mapping(name: str) -> Optional[Dict]:
-    edge_table = graph_edge_table(workspace, graph)
 
-    return {"nodeTables": node_tables, "edgeTable": edge_table}
+def search_user(query: str) -> Cursor:
```
Why can't this be a static method of the `User` class?

AlmightyYakob replied:
The main reason I didn't include it in that class is that it doesn't really operate on a specific user, or interact with the `User` class in anyway, like most of the other static methods do. I understand that it would be good to have for organization though, so I can move it.

waxlamp commented on multinet/db.py:
```diff
@@ -132,358 +107,33 @@ def workspace_mapping(name: str) -> Optional[Dict]:
-      FOR d in c
-        COLLECT WITH COUNT INTO count
-        RETURN count
+          LIMIT 50
```
I know it has nothing to do with this PR, but let's file an issue to make this search function pageable.

waxlamp replied:
#462

waxlamp commented on multinet/db.py:
```diff
@@ -500,118 +150,7 @@ def _run_aql_query(
-    }
-
-
+# TODO: Refactor the below functions into an `Upload` class
```
File an issue for this please, and link it in the comment.

waxlamp commented on multinet/util.py:
```diff
@@ -2,21 +2,81 @@
+restricted_document_keys = {"_rev", "_id"}
+
+
+# TODO: Remove once permission storage is updated
```
File/link issue here please.

waxlamp commented on multinet/util.py:
```diff
@@ -2,21 +2,81 @@
+    return new_permissions
+
+
+# TODO: Remove once permission storage is updated
```
Link issue here.

waxlamp commented on multinet/api.py:
```diff
@@ -1,48 +1,43 @@
+    if user is not None:
+        return util.stream(user.available_workspaces())
+
+    return util.stream(Workspace.list_public())
```
The docstring says "available" but the method call is `list_public()`. Should it be `list_available()`, or are we conflating the meanings of purely public workspaces and all available workspaces?

AlmightyYakob replied:
If it gets to this line, it means the user isn't logged in, and thus the only available workspaces are the ones that are public.

waxlamp replied:
Ah, my reading comprehension skills failed.

A few comments in this block of code might be in order, to help guide readers as to what's going on.

waxlamp commented on mypy_stubs/arango/cursor.pyi:
```diff
@@ -1,3 +1,21 @@
+from typing import Any
 
-Cursor = Generator[Any, None, None]
+# Below is autogenerated with stubgen
```
Neat, what is `stubgen`?

AlmightyYakob replied:
It's a tool included with mypy that makes it easy to generate boilerplate stubs https://mypy.readthedocs.io/en/stable/stubgen.html#automatic-stub-generation-stubgen

waxlamp commented on mypy_stubs/arango/graph.pyi:
```diff
@@ -1,8 +1,11 @@
     def vertex_collections(self) -> List[str]: ...
     def edge_definitions(self) -> List[Dict]: ...
+    def vertex(
+        self, vertex: Any, rev: Optional[Any] = ..., check_rev: bool = ...
```
We can't do better than `Any` here?

AlmightyYakob replied:
We can, my bad

waxlamp reviewed (changes requested):
This is great work. I left a *lot* of comments. What I suggest is that non-critical requested changes be filed as issues (with a link to the comment), while more critical ones be fixed before merge. We can discuss in the comment threads what is critical vs. not.

The other change I would like to see: I think that ORM classes are related enough to what's left in db.py that they should go into their own subpackage together: `db.py` should move to `db/__init__.py`, and the ORM classes should be moved to `db/orm/`. Let me know what you think.

AlmightyYakob commented on multinet/table.py:
```diff
@@ -0,0 +1,109 @@
+        """
+        return self.handle.insert_many(rows)
+
+    # This may be a reason to split Table up into EdgeTable and NodeTable
```
This was mainly a question of if it makes sense to split the `Table` class into a `NodeTable` and `EdgeTable` class, which inherit from `Table`. All of the common functionality would be in `Table`, and then methods specific to edge/node tables (like the function below this line, which is specific to edge tables) would be placed in their respective classes. Right now I think it's just something to keep in mind, not something that needs to be addressed now. Either way, I could file an issue for it if you think it's appropriate.

waxlamp replied:
Thanks for the explanation.

I opened #459 to track this. I would just delete the comment:

```suggestion
```

AlmightyYakob commented on multinet/table.py:
```diff
@@ -0,0 +1,109 @@
+        """Return all the keys in a table."""
+        return self.handle.keys()
+
+    def row_fields(self, filter_keys: bool = False) -> List[str]:
```
`keys()` returns the keys of all the documents in the table (arangodb collection). That is, all of the `_key` fields on each document. This function returns the set of keys that are present in _each_ document. Technically speaking, it returns the keys on the first document it finds in the collection, but since we ensure a consistent schema, I think this is fine.

AlmightyYakob commented on multinet/table.py:
```diff
@@ -0,0 +1,109 @@
+        """Return all the keys in a table."""
+        return self.handle.keys()
+
+    def row_fields(self, filter_keys: bool = False) -> List[str]:
```
Actually, now that you've brought it up, I don't think we should even have an option for allowing the system arango keys (`_key`, `_id`, `_rev`), since they're redundant, and we know they'll always be there. I don't see what use case we'd need them for, or why we'd need to retrieve them from this function instead of just specifying them. If you're okay with it I can remove the option entirely.

Separately, if anyone has a better name than `row_fields` please lmk.

waxlamp replied:
Removing that option sounds great.

I think we should call these "column headers" (or just "headers"), in keeping with the tabular data theme.

AlmightyYakob commented on multinet/table.py:
```diff
@@ -0,0 +1,109 @@
+        keys = []
+        cur = self.handle.find({}, limit=1)
+
+        # TODO: Determine what should happen if cur is empty
```
Technically it's possible to create a table without uploading any data to it, although none of our uploaders do that.

waxlamp replied:
Ok. Then (like I said above) I think this approach is correct.

```suggestion
```

AlmightyYakob commented on multinet/table.py:
```diff
@@ -0,0 +1,109 @@
+        self.handle.rename(new_name)
+        self.name = new_name
+
+    def insert(self, rows: List[Dict]) -> List[Dict]:
+        """
+        Insert rows into this table.
+
+        Returns the metadata from the documents inserted.
+        """
+        return self.handle.insert_many(rows)
```
Good catch, should I fix that in this PR?

waxlamp replied:
No, let's fix it in a future PR aimed at #406. I just wanted to leave the breadcrumb here. I'll mark this conversation "resolved".

AlmightyYakob commented on multinet/table.py:
```diff
@@ -0,0 +1,109 @@
+        """
+        props = self.handle.properties()
+        if not props["edge"]:
+            raise InternalServerError(f"Table {self.name} is not an edge table.")
```
@JackWilb I see your point regarding it being a 400 error, but there's actually no way to reach this error through the client, except for an error with our classification/handling of the data internally. However, you could probably invoke this intentionally by calling the endpoints directly. Put it in a bit of a weird spot. Specifically, you could invoke this error if you were to call [this](https://api.multinet.app/apidocs/#/graph/post_api_workspaces__workspace__graphs__graph_) endpoint with a node_table in the `edge_table` parameter. In that case, I don't think it should be a `RuntimeError`, since it is the user's fault at that point. This becomes more of an issue when other collaborators are hitting our API directly, instead of just through the client.

For the reasons I listed, I agree with Jack and think it should be a 400 level error. Curious what you think @waxlamp.

waxlamp replied:
Good argument for 4xx.

I guess it could be a [422](https://httpstatuses.com/422) but tbh just a 400 with an appropriate message (such as the one you've got here) would be fine.

AlmightyYakob replied:
47ec4bd

AlmightyYakob commented on multinet/graph.py:
```diff
@@ -0,0 +1,119 @@
+        """Return nodes in this graph."""
+        coll_names = self.handle.vertex_collections()
+
+        result_set_size = 0
+        query_size = 0
+        total_nodes = []
+
+        for coll_name in coll_names:
+            coll = self.handle.vertex_collection(coll_name)
+            result_set_size += coll.all().count()
+
+            if limit is None or query_size < limit:
+                remaining_limit = min(limit, limit - query_size) if limit else None
+                cur = coll.all(skip=offset, limit=remaining_limit)
+
+                query_size += cur.count()
+                total_nodes.extend(list(cur))
```
I tried using functions that already exist in `python-arango` as I think it makes sense if possible. However if it causes a significant reduction in performance compared to an AQL query, then obviously we should use that instead.

waxlamp replied:
:+1:

AlmightyYakob commented on multinet/table.py:
```diff
@@ -0,0 +1,109 @@
+
+        return {"count": count, "rows": list(rows)}
+
+    def row(self, doc: Union[Dict, str]) -> Union[Dict, None]:
```
If it's a `str`, it can be the document ID or key. https://python-driver-for-arangodb.readthedocs.io/en/master/specs.html#arango.collection.StandardCollection.get

waxlamp replied:
Great, thanks for the explanation.

BTW I noticed we can use `Optional` here as well:

```suggestion
    def row(self, doc: Union[Dict, str]) -> Optional[Dict]:
```

AlmightyYakob commented on multinet/db.py:
```diff
@@ -118,7 +118,7 @@ def workspace_mapping_collection() -> StandardCollection:
 # Caches the document that maps an external workspace name to it's internal one
 @lru_cache()
-def workspace_mapping(name: str) -> Optional[WorkspaceDocument]:
+def workspace_mapping(name: str) -> Optional[Dict]:
```
I tried to avoid dealing with `TypedDict`s in the first pass, since that was where much of the pain points came from originally. If we can deal with them in a sane way I'm willing to add them, but I think we need to come up with a methodology for using them.

waxlamp replied:
Let's worry about this in a followup. I can take it on myself: #460.

AlmightyYakob commented on multinet/workspace.py:
```diff
@@ -0,0 +1,356 @@
+class WorkspacePermissions:
+    """The permissions on a workspace."""
+
+    # TODO: Change str to User once updating permissions storage
```
Here's the issue https://github.com/multinet-app/multinet-server/issues/456. I can link it in the comment.

AlmightyYakob commented on multinet/workspace.py:
```diff
@@ -0,0 +1,356 @@
+
+    # TODO: Change str to User once updating permissions storage
+    owner: str
+    maintainers: List[str] = dataclasses.field(default_factory=lambda: [])
```
It's what's required when setting defaults on dataclass fields that aren't primitives.

AlmightyYakob commented on multinet/workspace.py:
```diff
@@ -0,0 +1,356 @@
+        if Workspace.exists(name):
+            raise AlreadyExists("Workspace", name)
+
+        internal = util.generate_arango_workspace_name()
+
+        try:
+            system_db(readonly=False).create_database(internal)
+        except DatabaseCreateError:
+            # Could only happen if there's a name collision
+            raise InternalServerError("Error creating workspace")
```
We could do that, however we probably wouldn't want to introduce unbounded while loops in our code base, so we would probably only do this a finite number of times before defaulting to raising this error. You could argue that based on [this](https://github.com/multinet-app/multinet-server/pull/358#discussion_r410721018) comment, the difference between the chance of this happening, and that chance multiplied by a finite number, is small enough to not worry about. That is, as soon as we start to worry about this chance, it's not clear where you'd draw the line.

waxlamp replied:
Right. Let's just do nothing.

AlmightyYakob commented on multinet/workspace.py:
```diff
@@ -0,0 +1,356 @@
+            "permissions": permissions,
+        }
+
+        coll = workspace_mapping_collection(readonly=False)
+        coll.insert(workspace_dict, sync=True)
+        workspace_mapping.cache_clear()
```
We could, but I didn't remove it here because I think if we're going to be possibly impacting the performance of the application, it makes sense to do that in it's own PR, so that the effects are easier to observe.

AlmightyYakob replied:
Related to #453

waxlamp replied:
Great. I forgot we were talking about using `memcached`!

AlmightyYakob commented on multinet/workspace.py:
```diff
@@ -0,0 +1,356 @@
+        return copy.deepcopy(doc)
+
+    # Graphs
+    def graphs(self) -> List[Dict]:
```
Same response as my other comment regarding TypedDicts

waxlamp replied:
#460

waxlamp commented on multinet/graph.py:
```diff
@@ -0,0 +1,119 @@
+        """Return the edge table of this graph."""
+        edge_defs = self.handle.edge_definitions()
+
+        # Currently assume one edge definition per graph
```
Something like:

```suggestion
        # Multinet graphs have just one edge definition.
```

waxlamp commented on multinet/table.py:
```diff
@@ -0,0 +1,109 @@
+    def row_fields(self, filter_keys: bool = False) -> List[str]:
+        """Return a the fields present on each row in this table."""
+        keys = []
+        cur = self.handle.find({}, limit=1)
```
@JackWilb if this satisfies you, please resolve the conversation thread.

jjnesbitt committed:
row_fields -> headers, remove filter_keys param

jjnesbitt committed:
Remove redundant code in Workspace.from_dict

jjnesbitt committed:
Fix Workspace.save, use in set_permissions

AlmightyYakob commented on multinet/workspace.py:
```diff
@@ -0,0 +1,356 @@
+        """Create a Workspace Object."""
+        self.name = name
+
+        # TODO: Don't access database right away
```
#463

jjnesbitt committed:
Update docstrings

jjnesbitt committed:
Fix comments

jjnesbitt committed:
Apply suggestions from code review

Co-authored-by: Roni Choudhury <2903332+waxlamp@users.noreply.github.com>
Co-authored-by: Jack Wilburn <jackwilburn@tutanota.com>

jjnesbitt committed:
Misc fixes/changes

- Move user search into User class
- Inline permissions assignment in workspace create
- Fix Workspace permissions use
- Remove staggered use of Workspace().permissions, and get_permissions.
- Add readonly flag to uploads_database
- Remove unnecessary cast to bool in `is_reader`
- Remove extra call when getting table rows in csv downloader
- Narrow type of arango Graph.vertex stub
- Add extra elif to node_edges func
- Remove immediate flag in db.db()

AlmightyYakob commented on multinet/workspace.py:
```diff
@@ -0,0 +1,356 @@
+        doc = self.get_metadata()
+
+        self.permissions = self.permissions or doc["permissions"]
+        self.internal = self.internal or doc["internal"]
```
This was actually due to a misuse of the Workspace class. I've removed the `or` expressions in 1995a03

jjnesbitt committed:
Create error for improper supplied table type

waxlamp reviewed (changes requested):
All the changes look good. Did you see my note in my first review about moving the ORM class files into an `orm` package within `db`? That's the last thing I'd request before we merge.

Really great work on this!

AlmightyYakob commented:
> All the changes look good. Did you see my note in my first review about moving the ORM class files into an `orm` package within `db`? That's the last thing I'd request before we merge.
> 
> Really great work on this!

Yes I recall that now. What do you think about keeping the `db.py` file where it is, and placing the class files into a `models` folder? That way, it's fairly explicit, and leaves just primitive database operations in the `db.py` file.

waxlamp commented:
> > All the changes look good. Did you see my note in my first review about moving the ORM class files into an `orm` package within `db`? That's the last thing I'd request before we merge.
> > Really great work on this!
> 
> Yes I recall that now. What do you think about keeping the `db.py` file where it is, and placing the class files into a `models` folder? That way, it's fairly explicit, and leaves just primitive database operations in the `db.py` file.

The name `model` is probably better than `orm`. But I still think it should be `db.model.Graph` etc. If we move `db.py` to `db/__init__.py`, and put all the class files in `db/model/`, then the import statements for the primitive operations formerly in `db.py` wouldn't have to change, right?

I think I'd prefer that solution, to make it all the way explicit that `db.model.Graph` is a database model object.

jjnesbitt committed:
Create db folder, place models in db/models/

JackWilb reviewed (approved):
Looks good to me! Tested creating/deleting workspaces, graphs, and tables, viewing nodes, edges, pagination, graphs, visualizing data locally. Appears to all be working.

waxlamp reviewed (approved):
Looks great. Thanks for running this marathon, @AlmightyYakob!
